tests/phpunit/integration/Core/Util/REST_Entity_Search_ControllerTest.php

```diff
@@ -0,0 +1,157 @@
+<?php
+/**
+ * REST_Entity_Search_ControllerTest
+ *
+ * @package   Google\Site_Kit\Tests\Core\Util
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Tests\Core\Util;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\REST_API\REST_Routes;
+use Google\Site_Kit\Core\Util\REST_Entity_Search_Controller;
+use Google\Site_Kit\Tests\TestCase;
+use WP_REST_Request;
+use WP_REST_Server;
+
+class REST_Entity_Search_ControllerTest extends TestCase {
+
+	/**
+	 * Controller instance.
+	 *
+	 * @var REST_Entity_Search_Controller
+	 */
+	private $controller;
+
+	public function setUp() {
+		parent::setUp();
+
+		$context          = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$this->controller = new REST_Entity_Search_Controller( $context );
+	}
+
+	public function tearDown() {
+		parent::tearDown();
+		// This ensures the REST server is initialized fresh for each test using it.
+		unset( $GLOBALS['wp_rest_server'] );
+		unset( $GLOBALS['current_user'] );
+	}
+
+	public function test_register() {
+		remove_all_filters( 'googlesitekit_rest_routes' );
+
+		$this->controller->register();
+
+		$this->assertTrue( has_filter( 'googlesitekit_rest_routes' ) );
+	}
+
+	public function test_unauthorized_request() {
+		$this->controller->register();
+
+		$request = new WP_REST_Request( WP_REST_Server::READABLE, '/' . REST_Routes::REST_ROOT . '/core/search/data/entity-search' );
+		$request->set_query_params(
+			array(
+				'query' => 'foo',
+			)
+		);
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertNotEquals( 200, $response->get_status() );
+		$this->assertArrayHasKey( 'code', $response->get_data() );
+		$this->assertEquals( 'rest_forbidden', $response->get_data()['code'] );
+	}
+
+
```

Comment:
Please remove these two extra blank lines.

Reply:
As much as I love the idea, I feel like this is overkill as at that point we will be testing WP's core search functionality more than our API endpoint and see very little value in it. The main thing to test here is whether it returns post for general keyword and entity for direct links as that's our implementation. Isn't it right?

Reply:
I have fixed the empty line and added more assertions to make sure it returns the correct entity. What do you think?

Reply:
> As much as I love the idea, I feel like this is overkill as at that point we will be testing WP's core search functionality more than our API endpoint and see very little value in it. The main thing to test here is whether it returns post for general keyword and entity for direct links as that's our implementation. Isn't it right?

Not really, IMHO. As things stand we are not verifying that we are even doing a search - the backend might simply be returning all of the posts. The integration test is not aware that the underlying implementation is using the WP search API. It should treat the code it's testing as a bit of a black box and simply verify the desired behaviour.

Reply:
re. testing core wp functionality: maybe someday wp search will become too smart and consider hello and howdy similar word and return howdy results as well for hello keyword, so that will break the test although It's not really broken.

Reply:
Again, I don't think this is a case where we should be considering what is core WP functionality. We should just be writing a test that verifies the endpoint behaviour. Besides, if the underlying search behaviour were to change like it, it would be a breaking change and as such I'd expect to see some test failures.

Reply:
Maybe we should ask for a second opinion here?

Reply:
> How about, rather than creating these posts/term inline in the tests, create them all in setUpBeforeClass.

IMO, that's fine to keep those posts and terms creation in test methods since they are needed only there. If we move them to the `setUpBeforeClass` method, all those posts and terms will be created for each test in this file, which is redundant.

> Also add a couple more which won't be matched.

Yes, it would be nice to have at least one more entity that doesn't match to make sure we get the correct number of entities in the response.

> we will be testing WP's core search functionality more than our API endpoint and see very little value in it.

Although we use core functions in the handler, we don't test WP's core search functionality here. As Tom said, we should treat our handler as a black box and verify that it returns the correct results for a certain request. In this case, the correct result is the number of returned entities and the fact that every entity in the returned response contains correct properties (`id`/`url`/`title`/`type`).

> maybe someday wp search will become too smart and consider hello and howdy similar word and return howdy results as well for hello keyword, so that will break the test although It's not really broken.

That's absolutely fine and the reason why we add these tests. If something changes in the core and breaks our tests, then we will know about it immediately and will be able to react to those changes in the core.

Reply:
Thanks @eugene-manuilov, appreciate you chipping in. For me, the key thing here is to make the tests more robust, which adding some extra post/terms and making the assertions more explicit will cover. The use of `setUpBeforeClass` is more of an organisational issue and as such I don't have such a strong preference either way.

@kuasha420, based on the above, are you happy to make changes to add an extra post/term and also verify the entity data in the response?